### PR TITLE
[api-minor] Add an `ignoreErrors` parameter to `PDFPageProxy.{render, getTextContent, getOperatorList}` to allow e.g. rendering to continue even if there are errors (issue 6342, issue 3795, bug 1130815)

### PR DESCRIPTION
**Pull request by Snuffleupagus**
Other PDF readers, e.g. Adobe Reader and PDFium (in Chrome), will attempt to render as much of a page as possible even if there are errors present.
Currently we just bail as soon the first error is hit, which means that we'll usually not render anything in these cases and just display a blank page instead.
By adding opt-in parameters to the `PDFPageProxy.{render, getTextContent, getOperatorList}` API methods, we can thus improve our handling of corrupt PDF files and allow e.g. the default viewer to handle errors slightly more gracefully.

Fixes, inasmuch it's possible since the PDF files are corrupt, e.g. issue #6342, issue #3795, and [bug 1130815](https://bugzilla.mozilla.org/show_bug.cgi?id=1130815) (and probably others too).

*Note:* Slightly smaller diff with https://github.com/mozilla/pdf.js/pull/8176/files?w=1.

---
This change is [Reviewable](https://reviewable.io/reviews/mozilla/pdf.js/8176)

### CODE REVIEW

File: web/pdf_viewer.js [hunk 1]
```diff
@@ -144,6 +146,7 @@ var PDFViewer = (function pdfViewer() {
     this.renderInteractiveForms = options.renderInteractiveForms || false;
     this.enablePrintAutoRotate = options.enablePrintAutoRotate || false;
     this.renderer = options.renderer || RendererType.CANVAS;
+    this.apiIgnoreErrors = options.apiIgnoreErrors || false;
```
Comment (timvandermeij) [inline web/pdf_viewer.js]: It's `true` at https://github.com/mozilla/pdf.js/pull/8176/commits/8bbca5999ea2544c032dc2ab25d93e949fa2617a#diff-a0d82fd09027b92c6e7bed99dc735626 and at https://github.com/mozilla/pdf.js/pull/8176/commits/8bbca5999ea2544c032dc2ab25d93e949fa2617a#diff-2930ccbe813e302a92f28334307aa4f6, but `false` in other places like here, which is confusing. I personally think we can make `true` the default to recover as much as possible, like other readers and PDF.js in other places in the code.

**Review by timvandermeij — COMMENTED** [review]
I really like the idea here. I'm wondering if there is something we can do about the size of the patch, i.e., if we can reduce the amount of parameter passing a bit, but it may very well be that it's not really possible to address that. Other than that, it looks promising and this should definitely be reviewed more in-depth as I just quickly went over the code.

File: web/pdf_viewer.js (hunk 1 above)
Comment (Snuffleupagus) [inline web/pdf_viewer.js]: Sure, there's no good reason why we cannot default to `true` everywhere in the viewer code; don't know why I didn't do that to be honest :-)

**Comment by Snuffleupagus** [comment]
> I really like the idea here.

Thank you :-)

> [...] if we can reduce the amount of parameter passing a bit, but it may very well be that it's not really possible to address that.

I had the same reaction *initially* while writing the code, but unfortunately I'm not sure if that's really possible, given how the code in `src/core/evaluator.js` is structured.
What could possibly help though, is to re-factor some of the method signatures to take parameter `Object`s rather than a bunch of parameters that you're just never sure in which order they should be!?

**Comment by timvandermeij** [comment]
Good point! I think that could help in some cases and in general it may be cleaner than passing ordered parameters.

**Review by yurydelendik — APPROVED** [review]
Looks good. But I would like to see ignoreErrors == true by default -- one thing less for users to define. Can we rename it to stopAtErrors/enableErrors at api facing code perhaps (core code can use ignoreErrors)?

**Comment by Snuffleupagus** [comment]
> But I would like to see ignoreErrors == true by default -- one thing less for users to define. Can we rename it to stopAtErrors/enableErrors at api facing code perhaps (core code can use ignoreErrors)?

Really good idea, I've made the suggested changes.
Thank you for the review!

**Comment by Snuffleupagus** [comment]
Closing in favour of PR #8240, since I wanted to preserve this branch in case the "new" approach isn't exactly what was asked for.

Notable changes in that PR:
 - Add a `stopAtErrors` parameter, defaulting to `true` at the API side; hence an API consumer will get this "recovery" mode by default now.
 - With the parameter now being at the `getDocument` level, the code becomes considerably simpler since the `evaluatorOptions` are used instead of passing a single parameter around to a whole bunch of methods in `src/core/evaluator.js`. Though, for simplicity with respect to both the code and e.g. existing unit-tests, the `src/core/` files still has the old behaviour by default.
 - Remove the (viewer) preferences, since I cannot imaging why anyone would want to turn this off in the default viewer,
 - Always ignore errors when parsing Type3 fonts, since I cannot see why it'd be clever to potentially render *partial* font glyphs.

**Pull request closed without merging**